### PR TITLE
Add unicode support to pscm

### PR DESCRIPTION
**Pull request by jingkaimori**
## What is changing

- Introduce icu4c
- Change inner representation of string from `std::string` to `icu::UnicodeString`
- Change logging system to log with Unicode
- Iterate over string with `CharacterIterator`
- Adopt parser to consume Unicode string
- Detect character property with icu4c instead of std
- Output objects and Unicode strings in doctest
- Run integrate tests by xmake
- Locale independent Number stringlizer and parser, `std::to_chars` should be polyfilled on Apple clang and older gcc
- **Disable** building via cmake on windows due to icu4c.

## What is missing

- [ ] ICU4C in ios, android and wasm, on cmake and bazel
- [ ] Format message with icu and remove fmt and spdlog library
- [ ] Miscellaneous error fix
- [ ] R6RS tests for Unicode string
- [ ] Optimization

### CODE REVIEW

File: include/pscm/logger/Event.h
Comment (PikachuHy) [inline include/pscm/logger/Event.h:1]: the Event.h and Logger.h are removed
I thank it is a mistake.

File: include/pscm/logger/Logger.hpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  static Logger *root_logger();
+  static std::unordered_map<std::string, Logger *>& logger_map();
+  static Logger *get_logger(std::string name);
+  void log(Logger::Level level, UString msg, SourceLocation loc = {});
+  void add_appender(Appender *appender);
+  template<typename... MsgT>
+  void log(Logger::Level level, const UString format_, SourceLocation loc, MsgT... msg)
+  {
+  if (is_level_enabled(level)) {
+    Event event;
```
Comment (PikachuHy) [inline include/pscm/logger/Logger.hpp]: error: variable has incomplete type 'Event'

**Comment by PikachuHy** [comment]
make sure pass the Clang Format Diff CI and all xmake CI

File: src/Parser.cpp [hunk 2]
```diff
@@ -718,77 +730,66 @@ Parser::Token Parser::next_token() {
...
-  }
+// 观览器 #4 的辅助类型
+template<class ... Ts>
+struct overloaded : Ts... { using Ts::operator()...; };
+
+/**
+ * 读取当前字符并前移迭代器，类似于 nextposinc 系列。返回EOF表示读到末尾。
+*/
+UChar32 Parser::next_char() {
+  UChar32 ch = std::visit(overloaded{
```
Comment (PikachuHy) [inline src/Parser.cpp:798]: src/Parser.cpp:750:27: error: no viable constructor or deduction guide for deduction of template arguments of 'overloaded'

File: CMakeLists.txt [hunk 3]
```diff
@@ -21,6 +21,33 @@ endif ()
...
+          )
+          message(INFO "Imported ICU::${component}")
+else()
+  message(WARNING "icu${component} not found")
+endif()
+endif()
+endfunction(_import_icu_)
+_import_icu_(uc)
+_import_icu_(i18n)
+_import_icu_(io)
```
Comment (PikachuHy) [inline CMakeLists.txt]: add `target_link_libraries(pscm PUBLIC ICU::uc ICU::i18n ICU::io)` to use icu
after declare target `add_library(pscm)`

**Comment by PikachuHy** [comment]
compile failed with clang16

```
#include <variant>
template <typename... Ts>
struct overloaded : Ts... {
  using Ts::operator()...;
};

int main(){
    std::variant<int, double> t;
    std::visit(
        overloaded{
            [](int t){},
            [](double t){}
        },
        t
    );
    return 0;
}
```

anything wrong?

**Comment by PikachuHy** [comment]
> compile failed with clang16
> 
> ```
> #include <variant>
> template <typename... Ts>
> struct overloaded : Ts... {
>   using Ts::operator()...;
> };
> 
> int main(){
>     std::variant<int, double> t;
>     std::visit(
>         overloaded{
>             [](int t){},
>             [](double t){}
>         },
>         t
>     );
>     return 0;
> }
> ```
> 
> anything wrong?

this one work fine

```

#include <variant>
template <typename... Ts> struct overloaded : Ts... {
  overloaded(Ts const &...ts) : Ts{ts}... {}
  using Ts::operator()...;
};

int main() {
  std::variant<int, double> t;
  std::visit(overloaded{[](int t) {}, [](double t) {}}, t);
  return 0;
}

```

File: src/misc/ICUCompat.cpp [hunk 4]
```diff
@@ -0,0 +1,155 @@
+//
+// Created by jingkaimori on 2023/8/16.
+//
+
+#pragma once
```
Comment (PikachuHy) [inline src/misc/ICUCompat.cpp]:
```
warning: #pragma once in main file [-Wpragma-once-outside-header]
```

only use `#pragma once` in headers

File: android/app/src/main/cpp/native-lib.cpp [hunk 5]
```diff
@@ -19,10 +19,12 @@ extern "C" JNIEXPORT jstring JNICALL Java_dev_pscm_android_MainActivity_evalSche
                                                                                        ,
                                                                                        jlong scm, jstring code) {
   auto _scm = (pscm::Scheme *)scm;
-  auto c_str = env->GetStringChars(code, nullptr);
```
Comment (jingkaimori) [inline android/app/src/main/cpp/native-lib.cpp]: 为什么只能用u8的字符串？icu4c的u16字符串应该兼容Java的字符串

Comment (PikachuHy) [reply android/app/src/main/cpp/native-lib.cpp]: build error

```
Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
android/app/src/main/cpp/native-lib.cpp:27:25: error: cannot initialize a parameter of type 'const jchar *' (aka 'const unsigned short *') with an rvalue of type 'const char16_t *'
  return env->NewString(s3.getBuffer(), s3.length());
                        ^~~~~~~~~~~~~~
external/androidndk/toolchains/llvm/prebuilt/darwin-x86_64/bin/../sysroot/usr/include/jni.h:827:36: note: passing argument to parameter 'unicodeChars' here
    jstring NewString(const jchar* unicodeChars, jsize len)
                                   ^
1 error generated.
Target //android/app/src/main:app failed to build

```

we will fix these error with a new PR.

**Pull request merged**